File: api-net/Controllers/ProductController.cs
using LiftOtonom.Api.Data;
using LiftOtonom.Api.Models;
using LiftOtonom.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace LiftOtonom.Api.Controllers;

[ApiController]
[Authorize]
[Route("api/v1/inventory")]
public class ProductController(AppDbContext db) : ControllerBase
{
    public record ProductDto(long? SupplierId, string? Code, string Name, string? Category, string? Unit,
        decimal? StockQuantity, decimal? MinStock, decimal? UnitPrice);
    public record MoveDto(decimal Quantity, decimal? UnitPrice, string? Note);

    [HttpGet]
    public async Task<IActionResult> Index([FromQuery] string? search, [FromQuery(Name = "low_stock")] bool lowStock = false,
        [FromQuery(Name = "per_page")] int perPage = 25, [FromQuery] int page = 1)
    {
        var q = db.Products.AsQueryable();
        if (!string.IsNullOrWhiteSpace(search))
            q = q.Where(p => EF.Functions.ILike(p.Name, $"%{search}%") ||
                             EF.Functions.ILike(p.Code ?? "", $"%{search}%") ||
                             EF.Functions.ILike(p.Category ?? "", $"%{search}%"));
        if (lowStock) q = q.Where(p => p.StockQuantity < p.MinStock);
        return Ok(await q.OrderBy(p => p.Name).ToPagedAsync(page, perPage));
    }

    [HttpGet("low-stock")]
    public async Task<IActionResult> LowStock() =>
        Ok(await db.Products.Where(p => p.StockQuantity < p.MinStock).OrderBy(p => p.Name).ToListAsync());

    [HttpGet("movements")]
    public async Task<IActionResult> Movements([FromQuery(Name = "product_id")] long? productId,
        [FromQuery(Name = "per_page")] int perPage = 25, [FromQuery] int page = 1)
    {
        var q = db.StockMovements.AsQueryable();
        if (productId is { } p) q = q.Where(m => m.ProductId == p);
        return Ok(await q.OrderByDescending(m => m.Id).ToPagedAsync(page, perPage));
    }

    [HttpPost]
    public async Task<IActionResult> Store(ProductDto dto)
    {
        var now = DateTime.UtcNow;
        var p = new Product
        {
            TenantId = db.CurrentTenantId!.Value, SupplierId = dto.SupplierId, Code = dto.Code, Name = dto.Name,
            Category = dto.Category, Unit = dto.Unit, StockQuantity = dto.StockQuantity ?? 0,
            MinStock = dto.MinStock ?? 0, UnitPrice = dto.UnitPrice, CreatedAt = now, UpdatedAt = now,
        };
        db.Products.Add(p);
        await db.SaveChangesAsync();
        return StatusCode(201, p);
    }

    [HttpGet("{id:long}")]
    public async Task<IActionResult> Show(long id) =>
        Ok(await db.Products.Include(p => p.Supplier).FirstOrDefaultAsync(p => p.Id == id) ?? throw new ApiException(404, "Ürün bulunamadı."));

    [HttpPut("{id:long}")]
    public async Task<IActionResult> Update(long id, ProductDto dto)
    {
        var p = await db.Products.FirstOrDefaultAsync(x => x.Id == id) ?? throw new ApiException(404, "Ürün bulunamadı.");
        p.SupplierId = dto.SupplierId;
        if (!string.IsNullOrWhiteSpace(dto.Name)) p.Name = dto.Name;
        p.Category = dto.Category; p.Unit = dto.Unit;
        if (dto.MinStock is { } ms) p.MinStock = ms;
        p.UnitPrice = dto.UnitPrice; p.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync();
        return Ok(p);
    }

    [HttpDelete("{id:long}")]
    public async Task<IActionResult> Destroy(long id)
    {
        var p = await db.Products.FirstOrDefaultAsync(x => x.Id == id) ?? throw new ApiException(404, "Ürün bulunamadı.");
        p.DeletedAt = DateTime.UtcNow;
        await db.SaveChangesAsync();
        return Ok(new { message = "Ürün silindi." });
    }

    [HttpPost("{id:long}/stock-in")]
    public Task<IActionResult> StockIn(long id, MoveDto dto) => Move(id, dto, "in");

    [HttpPost("{id:long}/stock-out")]
    public Task<IActionResult> StockOut(long id, MoveDto dto) => Move(id, dto, "out");

    private async Task<IActionResult> Move(long id, MoveDto dto, string type)
    {
        if (dto.Quantity <= 0) throw new ApiException(422, "Miktar 0'dan büyük olmalı.");
        await using var tx = await db.Database.BeginTransactionAsync();
        var p = await db.Products.FirstOrDefaultAsync(x => x.Id == id) ?? throw new ApiException(404, "Ürün bulunamadı.");

        if (type == "out" && p.StockQuantity < dto.Quantity) throw new ApiException(422, "Yeterli stok yok.");
        p.StockQuantity += type == "in" ? dto.Quantity : -dto.Quantity;
        p.UpdatedAt = DateTime.UtcNow;

        var price = dto.UnitPrice ?? p.UnitPrice;
        db.StockMovements.Add(new StockMovement
        {
            TenantId = db.CurrentTenantId!.Value, ProductId = p.Id, Type = type, Quantity = dto.Quantity,
            UnitPrice = price, TotalPrice = dto.Quantity * (price ?? 0),
            ReferenceType = type == "in" ? "purchase" : "adjustment", Note = dto.Note,
            CreatedBy = long.Parse(User.FindFirst("uid")!.Value), CreatedAt = DateTime.UtcNow,
        });
        await db.SaveChangesAsync();
        await tx.CommitAsync();
        return Ok(new { message = "Stok güncellendi.", stock_quantity = p.StockQuantity });
    }
}

File: api/app/Http/Controllers/Api/ProductController.php
<?php
// app/Http/Controllers/Api/ProductController.php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Models\Product;
use App\Models\StockMovement;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;

class ProductController extends Controller
{
    public function index(Request $request)
    {
        $q = Product::query()->with('supplier:id,name');
        if ($search = $request->string('search')->trim()->value()) {
            $q->where(fn ($w) => $w->where('name', 'ilike', "%{$search}%")
                ->orWhere('code', 'ilike', "%{$search}%")
                ->orWhere('category', 'ilike', "%{$search}%"));
        }
        if ($request->boolean('low_stock')) {
            $q->whereColumn('stock_quantity', '<', 'min_stock');
        }

        return $q->orderBy('name')->paginate(min((int) $request->integer('per_page', 25), 100));
    }

    public function store(Request $request)
    {
        $data = $request->validate([
            'supplier_id'    => ['nullable', 'integer', 'exists:suppliers,id'],
            'code'           => ['nullable', 'string', 'max:100'],
            'name'           => ['required', 'string', 'max:255'],
            'category'       => ['nullable', 'string', 'max:100'],
            'unit'           => ['nullable', 'string', 'max:20'],
            'stock_quantity' => ['nullable', 'numeric'],
            'min_stock'      => ['nullable', 'numeric'],
            'unit_price'     => ['nullable', 'numeric'],
        ]);

        return response()->json(Product::create($data)->fresh(), 201);
    }

    public function show(Product $product)
    {
        return $product->load('supplier:id,name');
    }

    public function update(Request $request, Product $product)
    {
        $product->update($request->validate([
            'supplier_id' => ['nullable', 'integer', 'exists:suppliers,id'],
            'name'        => ['sometimes', 'string', 'max:255'],
            'category'    => ['nullable', 'string', 'max:100'],
            'unit'        => ['nullable', 'string', 'max:20'],
            'min_stock'   => ['nullable', 'numeric'],
            'unit_price'  => ['nullable', 'numeric'],
        ]));

        return $product->fresh();
    }

    public function destroy(Product $product)
    {
        $product->delete();

        return response()->json(['message' => 'Ürün silindi.']);
    }

    public function stockIn(Request $request, Product $product)
    {
        return $this->move($request, $product, 'in');
    }

    public function stockOut(Request $request, Product $product)
    {
        return $this->move($request, $product, 'out');
    }

    public function lowStock()
    {
        return Product::whereColumn('stock_quantity', '<', 'min_stock')
            ->orderBy('name')->get();
    }

    public function movements(Request $request)
    {
        $q = StockMovement::query();
        if ($request->filled('product_id')) {
            $q->where('product_id', $request->integer('product_id'));
        }

        return $q->orderByDesc('id')->paginate(min((int) $request->integer('per_page', 25), 100));
    }

    private function move(Request $request, Product $product, string $type)
    {
        $data = $request->validate([
            'quantity'   => ['required', 'numeric', 'min:0.01'],
            'unit_price' => ['nullable', 'numeric'],
            'note'       => ['nullable', 'string'],
        ]);

        return DB::transaction(function () use ($product, $type, $data, $request) {
            $product = Product::lockForUpdate()->find($product->id);
            $qty = (float) $data['quantity'];

            if ($type === 'out' && (float) $product->stock_quantity < $qty) {
                abort(422, 'Yeterli stok yok.');
            }

            $product->stock_quantity = (float) $product->stock_quantity + ($type === 'in' ? $qty : -$qty);
            $product->save();

            StockMovement::create([
                'product_id'  => $product->id,
                'type'        => $type,
                'quantity'    => $qty,
                'unit_price'  => $data['unit_price'] ?? $product->unit_price,
                'total_price' => $qty * (float) ($data['unit_price'] ?? $product->unit_price),
                'reference_type' => $type === 'in' ? 'purchase' : 'adjustment',
                'note'        => $data['note'] ?? null,
                'created_by'  => $request->user()->id,
                'created_at'  => now(),
            ]);

            return response()->json([
                'message' => 'Stok güncellendi.',
                'stock_quantity' => $product->stock_quantity,
            ]);
        });
    }
}
